samples/core/enumopencl/main.c

```diff
@@ -0,0 +1,243 @@
+}
+
+static cl_int PrintDeviceInfoSummary(cl_device_id* devices, cl_uint numDevices)
+{
+    cl_int errorCode = CL_SUCCESS;
+
+    cl_device_type deviceType;
+    char* deviceName = NULL;
+    char* deviceVendor = NULL;
+    char* deviceVersion = NULL;
+    char* deviceProfile = NULL;
+    char* driverVersion = NULL;
+
+    cl_uint i = 0;
+    for (i = 0; i < numDevices; i++)
+    {
+        errorCode |= clGetDeviceInfo(devices[i], CL_DEVICE_TYPE,
+                                     sizeof(deviceType), &deviceType, NULL);
+        errorCode |= AllocateAndGetDeviceInfoString(devices[i], CL_DEVICE_NAME,
+                                                    &deviceName);
+        errorCode |= AllocateAndGetDeviceInfoString(
+            devices[i], CL_DEVICE_VENDOR, &deviceVendor);
+        errorCode |= AllocateAndGetDeviceInfoString(
+            devices[i], CL_DEVICE_VERSION, &deviceVersion);
+        errorCode |= AllocateAndGetDeviceInfoString(
+            devices[i], CL_DEVICE_PROFILE, &deviceProfile);
+        errorCode |= AllocateAndGetDeviceInfoString(
+            devices[i], CL_DRIVER_VERSION, &driverVersion);
+
+        if (errorCode == CL_SUCCESS)
+        {
+            printf("Device[%u]:\n", i);
+
+            PrintDeviceType("\tType:           ", deviceType);
+
+            printf("\tName:           %s\n", deviceName);
+            printf("\tVendor:         %s\n", deviceVendor);
+            printf("\tDevice Version: %s\n", deviceVersion);
+            printf("\tDevice Profile: %s\n", deviceProfile);
+            printf("\tDriver Version: %s\n", driverVersion);
+        }
+        else
+        {
+            fprintf(stderr, "Error getting device info for device %u.\n", i);
+        }
+
+        free(deviceName);
+        free(deviceVendor);
+        free(deviceVersion);
+        free(deviceProfile);
+        free(driverVersion);
+
+        deviceName = NULL;
+        deviceVendor = NULL;
+        deviceVersion = NULL;
+        deviceProfile = NULL;
+        driverVersion = NULL;
+    }
+
+    return errorCode;
+}
+
+int main(int argc, char** argv)
+{
+    cl_uint numPlatforms = 0;
+    clGetPlatformIDs(0, NULL, &numPlatforms);
+    printf("Enumerated %u platforms.\n\n", numPlatforms);
+
+    cl_platform_id* platforms =
+        (cl_platform_id*)malloc(numPlatforms * sizeof(cl_platform_id));
+    clGetPlatformIDs(numPlatforms, platforms, NULL);
+
+    cl_uint i = 0;
+    for (i = 0; i < numPlatforms; i++)
+    {
+        printf("Platform[%u]:\n", i);
+        PrintPlatformInfoSummary(platforms[i]);
+
+        cl_uint numDevices = 0;
+        clGetDeviceIDs(platforms[i], CL_DEVICE_TYPE_ALL, 0, NULL, &numDevices);
+
+        cl_device_id* devices =
+            (cl_device_id*)malloc(numDevices * sizeof(cl_device_id));
+        clGetDeviceIDs(platforms[i], CL_DEVICE_TYPE_ALL, numDevices, devices,
+                       NULL);
+
+        PrintDeviceInfoSummary(devices, numDevices);
+        printf("\n");
+
+        free(devices);
```

Comment:
Even though devices obtained by `clGetDeviceIDs` will return "root-level" devices which never get freed, I would highly suggest promoting "brainlessly" calling `clReleaseDevice()` on anything that was created at scope-level and then the scope closes without transferring ownership of those devices.

It's just a hygenic thing, much like freeing memory right before main exits. It's not strictly required, the OS will clean up anyway, but the reflex should develop in people to control the lifetime of _all_ OpenCL objects that have retain/release semantics.

Reply:
I think I'd prefer not to do this.  Here's why:

- Most uses of device handles will not create sub-devices, so most uses will not need to release device handles.  I'd prefer not to put extranenous code in such a simple sample.
- clReleaseDevice is an OpenCL 1.2 function, so it may behave unpredictably for an OpenCL 1.0 or 1.1 implementation.

Are you OK leaving this as-is?

Reply:
Okay, this isn't the hill I'd want to die on. I'd air on the side of building fail-safe muscle reflexes instead of inviting discussion on why omitting this is valid in this particular case.